include/shark/Models/Trees/General.h

```diff
@@ -0,0 +1,126 @@
+//===========================================================================
+/*!
+ *
+ *
+ * \brief       General functions for Tree modeling.
+ *
+ *
+ *
+ * \author      J. Wrigley
+ * \date        2016
+ *
+ *
+ * \par Copyright 1995-2015 Shark Development Team
+ *
+ * <BR><HR>
+ * This file is part of Shark.
+ * <http://image.diku.dk/shark/>
+ *
+ * Shark is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Shark is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Shark.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+//===========================================================================
+
+#ifndef SHARK_MODELS_TREES_GENERAL_H
+#define SHARK_MODELS_TREES_GENERAL_H
+#include <memory>
+template<class T> class TD;
+#define typeof(param) TD<decltype(param)>{}
+namespace shark {
+namespace detail {
+namespace cart {
+// Helper functions
+template<class T> using sink = T;
```

Comment:
When compiled `sink<T&>` evaluates to `T&`. `sink<T&>` is used to mark parameters that will be moved from eventually, and whose state therefore will be undefined subsequently.

It solves 2 problems:
1. It informs the user that we intend to invalidate the object later by moving from it. The alternative way to do this would be to make the function take a `T&&` and call it `func(std::move(arg))`, but this doesn't solve the second problem.
2. In cases when move can't actually be performed (say the object being moved doesn't implement move semantics (yet)), it is faster to use `sink<T&>` going down the call-chain and then `std::move`, rather than to `std::move` on every call down, because in those cases move would then become copy.

sink & move thus evaluates to reference & copy, rather than n*copy in situations when move-semantics aren't fully supported.